Implement CLI cmd to run file into console

Isalafont opened the pull request:
Add convertJson2CSV_cmd file
Write a script to run program into console with passing 2 ARGV.
Transform module ConvertJson2CSV into class
Still need to pass arguments when calling file in the console !!

Add a different json_file to test the lib.
generate empty CSV.

TODO
Find a way to launch program using only the cmd 'ruby convertJson2CSV_cmd.rb'
Fix the add_values method to fetch the values and adding it to CSV
Change recursive method to flatten keys

Isalafont committed:
Implement CLI cmd to run file into console

Add convertJson2CSV_cmd file
Write a script to run program into console with passing 2 ARGV.
Transform module ConvertJson2CSV into class
Still need to pass arguments when calling file in the console !!

Add a different json_file to test the lib.
generate empty CSV.

TODO
Find a way to launch program using only the cmd 'ruby convertJson2CSV_cmd.rb'
Fix the add_values method to fetch the values and adding it to CSV
Change recursive method to flatten keys

JCBlondel commented on README.md:
````diff
@@ -9,7 +9,7 @@ The goal of this test is to write a small Ruby lib aiming to convert JSON files
 
 ```shell
-$ ruby lib/convertJsonCSV.rb
+$ ruby lib/convertJsonCSV_cmd.rb 'data_input/{#name_input_file}' 'data_output/{#name_output_file}
````
Directement lancer le programme, devrait te permettre de convertir tous les fichiers contenus dans /data_input

```suggestion
$ ruby lib/convertJsonCSV_cmd.rb
```

JCBlondel commented on README.md:
```diff
@@ -39,8 +39,8 @@ Some sample files are located in `./input_json_files` and `./output_csv_files`
-- L'objectif global du test : Lorsque l'on execute la cmd "\$ruby le_programme_d'entrée.rb", tous les fichiers _.json qui sont dans /input_json_files, soient convertis (ou non) en _.csv dans /output_csv_files.
-  _Les fichiers Input peuvent ne pas être sous format JSON ET/OU formatés différement._
+- L'objectif global du test : Lorsque l'on execute la cmd "\$ruby le*programme_d'entrée.rb", tous les fichiers *.json qui sont dans /input*json_files, soient convertis (ou non) en *.csv dans /output*csv_files.
+  \_Les fichiers Input peuvent ne pas être sous format JSON ET/OU formatés différement.*
```
Tu peux le tester au travers de tests Rspec si tu veux :) mais ce n'est pas un critère déterminant pour le test, il s'agit d'un exemple d'erreur qui puisse être gérée dans des tests unitaires

Isalafont replied:
J'ai commencé à chercher comment récupérer les les fichiers en utilisant 
Dir.children("#{FILE_PATH}")

def find_file
  file_list = Dir.children("#{FILE_PATH}")
end

Cela me permet de store un array avec les noms des fichiers ["users.json", "new_story.json"]

def select_json
  Pour chaque element de file_list
    sélectionner tous les elements se terminant par .json et l'enregistrer dans une variable
  Fin de la selection
end

Dig into Dir and File doc to select all the files from dir and convert it to csv.
case *json 
  => ouvrir chaque element avec JSON.parse(File.read(...))
case !*json 
   => File.read("....")

Je suis toujours en cours de reflexion pour cette partie du test.

JCBlondel commented on convertJsonCSV_cmd.rb:
```diff
@@ -0,0 +1,21 @@
+require_relative 'lib/convertJsonCSV'
+
+# Create programme to launch lib converter from CLI using name of ruby file only
+
+if __FILE__ == $0
+
+  # Method take 2 cmd line ARGV (input_file, output_file)
+  # Raise exception if there is not 2 arguments
+  
+  if ARGV.size != 2
+    raise('Wrong args : ruby convertJsonCSV_cmd.rb input.json output.csv')
+  end
+
+  convert = ConvertJsonCSV.new('./data_input/badges.json', './data_output/badges.csv')
+  convert.create_csv(ARGV[0], ARGV[1])
+end
+
+# Create cmd line argument : ARGV[0], ARGV[1] to replace input_file/output_file
+# => Running ruby convertJsonCSV_cmd.rb file.json file.csv
+
+# TODO find a way to only have to run cmd ruby convertJsonCSV_cmd.rb
```
Ce n'est pas nécessaire d'avoir un système de "prompt" pour sélectionner des fichiers à traiter par le programme :)

Isalafont replied:
Je viens de le retirer. 👍

JCBlondel commented on data_input/badges.json:
```diff
@@ -0,0 +1,30 @@
+{
```
Introduire d'autres fichiers .json pour éprouver ton programme, c'est super 👍 !

Isalafont replied:
Je me suis rendue compte avec d'autres fichiers que ma méthode create_header avait cassé. 
Je dois la refaire. Je réfléchie à utiliser la méthode `flatten` pour aplatir les headers mais je dois mieux gérer les conditions pour cette méthode recursive

JCBlondel commented on lib/convertJsonCSV.rb:
```diff
@@ -1,9 +1,19 @@
 require 'csv'
 
-module ConvertJsonCSV
+class ConvertJsonCSV
+
+  attr_reader :input_file
+  attr_accessor :output_file
```
👍 super lisible

Pars du postulat que tu n'as qu'un seul fichier pour atteindre l'objectif :)

JCBlondel commented on lib/convertJsonCSV.rb:
```diff
@@ -1,9 +1,19 @@
+
   
-  def self.create_csv(input_file, output_file)
+  def create_csv(input_file, output_file)
```
Si tu as des reader/accessor sur ta classe à l'initialisation, tu peux y accéder où que tu sois dans la classe, pas besoin de les passer en paramètre

Isalafont replied:
J'ai fait les modifications

JCBlondel commented on lib/convertJsonCSV.rb:
```diff
@@ -15,19 +25,19 @@ def self.create_csv(input_file, output_file)
   private
 
-  def self.get_data_from_json(input_file)
+  def get_data_from_json(input_file)
```
Idem, tu peux déjà accéder à input_file

JCBlondel commented on lib/convertJsonCSV.rb:
```diff
@@ -1,9 +1,19 @@
 require 'json'
 require 'csv'
```
Je te propose d'utiliser 'pry' pour placer des points d'arrêt dans le code lors de son execution :)

Est-ce que tu connais cet outils ? https://github.com/pry/pry

Pour arrêter l'exécution du code tu écris "binding.pry" et tu as alors accès au contexte en console, regarder les valeurs des variables, appeler des méthodes, tout

Isalafont replied:
Merci ! Je vais regarder cela et l'utiliser pour debugger mes méthodes :)

JCBlondel commented on lib/convertJsonCSV.rb:
```diff
@@ -15,19 +25,19 @@ def self.create_csv(input_file, output_file)
   end
 
-  def self.create_header(data_json)
+  def create_header(data_json)
```
De la même manière que tu stocks input_file et output_file à l'échelle de la classe, tu peux déclarer @data_json = [] dans la méthode #create_csv pour y accéder depuis n'importe quelle méthode

JCBlondel replied:
La récupération du header est une responsabilité assez forte, et avoir une classe dédiée, comporte d'énormes avantages

Isalafont replied:
Je modifie et corrige `@data_json` 
J'ai tenté de bouger la responsabilité de récupération du header et des rows dans une autre classe mais j'ai toujours des erreurs dans la méthode get_key que je n'arrive pas à débugger.

JCBlondel reviewed (commented):
👍

Isalafont committed:
CLI can call program just calling file

Add FILE_PATH for input and output
Try to keep ARGV[0] and ARGV[1] but create bug
Test new input json file => csv does not create
Failed to method create_header

Isalafont committed:
Test to lunch programe with Class LoadData

No methodError / undifined method load_data in create_csv

Isalafont committed:
Remove Prompt system to call program

Isalafont committed:
Remove parametre from create_csv and save_csv

Since input_file and output_file are accessible from attr_reader and attr_accessor
remove them from methods parameter

Isalafont committed:
call @data_json into create_csv

Modification in every method according to this change

Isalafont committed:
Try to create new class for create_header responsability

Create 2 new class
1 for loading Data
2 for create_header responsability

Since changing module to class / got difficulties to make code works again.
Stuck in the code to fix the issues

Create_header still not working / try binding.pry but did not find how to fix

Isalafont committed:
Rename get_header to header

Clean up code
Still can't get headers working
